active-rfcs/0000-build-performance.md

```diff
@@ -0,0 +1,187 @@
+- Start Date: 2021-12-13
+- Reference Issues:
+  - Previous: https://github.com/withastro/rfcs/pull/44
+- Implementation PR: https://github.com/withastro/astro/pull/2168
+
+# Summary
+
+In order to improve build performance and scale to sites that build thousands of pages, deprecate dynamic features that forces our build to scan rendered HTML and provide static replacements.
+
+These features will be required for SSR support as well, so this RFC seeks to unify how Astro sites are built in a way that is workable both for static-site generation (SSG) and server-side rendering (SSR).
+
+# Example
+
+The following features are bundled as part of this RFC:
+
+- Enforced static use of `client:` directives.
+- Deprecation of `Astro.resolve`.
+- Introduction of `local:` directive to act as a static replacement for Astro.resolve.
+
+## Static client hydration directives
+
+The directives used for client hydration are already static by implementation; if you attempt to add `client:load`, etc, during rendering Astro will not build that component's client-side JavaScript.
+
+This change will make this behavior defined and enforced. You can achieve dynamic usage by making the attribute static, but dynamically render a hydrated component like so:
+
+```astro
+---
+import Clock from '../components/Clock.jsx';
+const { shouldHydrate } = Astro.props;
+---
+{ shouldHydrate ? <Clock client:idle /> : <Clock /> }
+```
+
+Astro will be able to build the above correctly, but only JavaScript will be added to the page if the `shouldHydrate` condition is truthy.
+
+## Local directive
+
+In order to replace `Astro.resolve`, which can't be used to build things like styles and images, this RFC proposes a new sugar syntax:
+
+```astro
+<img local:src="../images/penguin.png" />
+```
+
+Note that this can already be achieved without sugar syntax by using an import statement:
+
+```astro
+---
+const imgUrl = '../images/penguin.png';
+---
+
+<img src={imgUrl} />
+```
+
+The sugar syntax is being introduced because we feel the above is harder to learn and less obvious of a solution.
+
+## Deprecation of Astro.resolve
+
+The `Astro.resolve` method was added to enable adding relative links to assets in the `src/` folder. However, since this is a function it can take dynamic values that are not compatible with a static build. For example:
+
+```astro
+---
+const { animal } = Astro.props;
+---
+<img src={Astro.resolve(`../images/${animal}.png`)} />
+```
+
+Instead, you can use `import.meta.glob` to dynamically add a URL:
+
+```astro
+---
+const images = import.meta.glob('../images/*.png');
+
+const { animal } = Astro.props;
+
+const animalUrl = Object.entries(images)
+.filter(([path]) => (
+  path.includes(animal)
+))
+.map(([, mod]) => mod.default)[0];
+---
+
+<img src={animalUrl} />
+```
+
+The above will result in *all* of the images in `../images/` getting built, but only the one you select will be used.
```

Comment:
If you're mentioning workarounds here, there are two that I think work today in Vite and are worth calling out here or somewhere else in the doc:

1. `const animalUrl = await import('../images/${Astro.props.animal}.png')`
2. "just put the image in `public/`"

If this is a look ahead at how we will document this, I'm a bigger fan of documenting dynamic import than `import.meta.glob`

Reply:
Didn't realize the dynamic import one could be built. Agree that's better to document and is less code, I'll replace the example with that.

And also note that `public/` is always an option.

Reply:
Double check the dynamic import if you can! I know Vite documents this support but it would be great to double check

Reply:
I've switched to dynamic import in the example.

Reply:
YES! I really like that. We could even go one step further and translate to:

```
<!-- INPUT -->
<img local:src={`../images/${animal}.png`} />

<!-- OUTPUT -->
<img src={await import(`../images/${animal}.png`)} />
```

Also it would be nice to flesh out this behavior in a bit more detail. For example, in that case I assume we still couldn't support `local:src={x}` or `local:src={'../' + 'image.png'}` and would warn on those in the compiler? Is that feasible based on how the compiler works? 

Also, what would trigger a dynamic import vs. a static ESM import? Or, would we always do dynamic import? If we always did a dynamic import, could we then rely on Vite to handle the "this `local:src` value isn't supported" behavior by warning/erroring when it sees the final `import()` expression?

Just some thoughts!

Reply:
Yeah, I think translating to always using the dynamic import is cleaner as far as implementation. That way we don't have to care or inspect what the value is, and you'd get an error if the usage is too dynamic for Vite to handle.

Reply:
🥳  yes!!! this one small feature would really powerful to include into astro, 
oh the possibilities......